File: Keyboard/KeyCommitQueue.swift
// KeyCommitQueue — thứ tự chốt phím khi gõ chồng ngón (research 25/09/2026).
//
// Phím chữ chèn lúc CHẠM XUỐNG; space / dấu câu / số-ký hiệu / return chèn lúc
// NHẤC. Gõ nhanh hai ngón cái ("anh␣em"): ngón trên space chưa nhấc thì "e" đã
// chèn → "anhe" rồi mới tới space — cảm giác "space không ăn", và "e" bị engine
// gộp vào từ cũ. Stock iOS chốt đúng thứ tự: phím đang đè được chốt ngay khi có
// ngón MỚI chạm xuống.
//
// Hàng đợi này giữ các phím nhấc-mới-chốt đang đè (theo thứ tự chạm):
//   • arm      — phím chạm xuống: ghi nhận, CHƯA chèn.
//   • flush    — một phím KHÁC chạm xuống: chốt mọi phím đang giữ, đúng thứ tự.
//   • release  — nhấc ngón HOẶC touch bị hệ thống huỷ: chốt nếu còn giữ (touch
//                bị cancel ở hàng dưới sát home indicator không còn nuốt phím).
//   • disarm   — bỏ không chốt (giữ space vào chế độ trackpad: nhả ra không có
//                dấu cách như stock).
// Pure, không UIKit — pinned by KeyCommitQueueTests.
final class KeyCommitQueue {
    private var pending: [(id: ObjectIdentifier, fire: () -> Void)] = []

    var isEmpty: Bool { pending.isEmpty }

    func arm(_ id: ObjectIdentifier, fire: @escaping () -> Void) {
        // Same key pressed again while the earlier press is still down (two thumbs
        // on space): the earlier press is a real keystroke — commit it, don't drop it.
        release(id)
        pending.append((id, fire))
    }

    /// Commit every pending key except `id`, in press order.
    func flush(except id: ObjectIdentifier? = nil) {
        guard !pending.isEmpty else { return }
        let fired = pending.filter { $0.id != id }
        pending.removeAll { $0.id != id }
        for p in fired { p.fire() }
    }

    /// Commit `id` if it is still pending (touch up / cancelled).
    func release(_ id: ObjectIdentifier) {
        guard let i = pending.firstIndex(where: { $0.id == id }) else { return }
        let p = pending.remove(at: i)
        p.fire()
    }

    /// Drop `id` without committing.
    func disarm(_ id: ObjectIdentifier) {
        pending.removeAll { $0.id == id }
    }
}

File: android/keyboard/src/main/kotlin/com/viettelex/keyboard/KeyCommitQueue.kt
package com.viettelex.keyboard

/**
 * Thứ tự chốt phím khi gõ chồng ngón (port iOS). Phím chữ chèn lúc DOWN; space /
 * dấu câu / số / return: [arm] lúc DOWN, [release] lúc UP hoặc CANCEL, [flush] khi
 * ngón KHÁC chạm xuống, [disarm] khi vào trackpad. [id] = định danh phím (so sánh
 * bằng identity: truyền cùng object phím).
 */
class KeyCommitQueue {
    private class Pending(val id: Any, val fire: () -> Unit)
    private val pending = ArrayList<Pending>(4)

    val isEmpty: Boolean get() = pending.isEmpty()

    fun arm(id: Any, fire: () -> Unit) {
        // cùng phím đè lần nữa khi lần trước chưa nhấc: lần trước là phím thật — chốt.
        release(id)
        pending.add(Pending(id, fire))
    }

    /** Chốt mọi phím đang chờ trừ [except], theo thứ tự chạm. */
    fun flush(except: Any? = null) {
        if (pending.isEmpty()) return
        val fired = pending.filter { it.id !== except }
        pending.removeAll { it.id !== except }
        for (p in fired) p.fire()
    }

    fun release(id: Any) {
        val i = pending.indexOfFirst { it.id === id }
        if (i < 0) return
        pending.removeAt(i).fire()
    }

    fun disarm(id: Any) { pending.removeAll { it.id === id } }
}
